controllers/userController.js

```diff
@@ -180,33 +180,112 @@ exports.RegisterVerify = async (req, res) => {
+            return res.status(401).send({ message: 'Tên đăng nhập hoặc mật khẩu không đúng' });
         }
+
         let user = result.recordset[0];
+
         let isPassword = await bcrypt.compare(password, user.PASSWORD);
         if (!isPassword) {
-            return res.status(400).send({ message: 'Mat khau khong hop le' });
+            return res.status(401).send({ message: 'Tên đăng nhập hoặc mật khẩu không đúng' });
         }
-        let token = jwt.sign({ userID: user.ID, userName: user.USERNAME, role: user.ROLE }, process.env.secretPass, { expiresIn: '1h' });
-        res.json({ message: `Dang nhap thanh cong với vai trò ${user.ROLE}`, token: token });
+        let token = jwt.sign(
+            { userID: user.ID, userName: user.USERNAME, role: user.ROLE },
+            process.env.secretPass,
+            { expiresIn: '2m' }
+        );
+        let refreshToken = jwt.sign(
+            { userID: user.ID, userName: user.USERNAME, role: user.ROLE },
+            process.env.secretPass,
+            { expiresIn: '7d' }
+        );
+
+        //hash refresh token
+        let hashReftoken = await bcrypt.hash(refreshToken, 10);
+        // delete old refresh token and insert new one into the database
+        await pool.request()
+            .input('hashtoken', hashReftoken)
+            .input('user_id', user.ID)
+            .input('expiresAt', expiresAt)
+            .query(
+                `DELETE REFRESH_TOKEN WHERE USER_ID = @user_id;
+                INSERT INTO REFRESH_TOKEN(TOKEN, USER_ID, EXPIRESAT) VALUES(@hashtoken, @user_id, @expiresAt)`
+            );
+        
+        res.json({ message: `Dang nhap thanh cong với vai trò ${user.ROLE}`, token: token , refreshToken: refreshToken});
 
     } catch (error) {
         console.error(error);
         res.status(500).send({ message: 'Khong the dang nhap' });
     }
 }
+exports.refreshToken = async (req, res) => {
+    const { refreshToken } = req.body;
+
+    try {
+        let pool = await connectionDB();
+        // Compare provided refresh token  with all tokens stored in table 
+        let check = await pool.request().query('SELECT *FROM REFRESH_TOKEN');
+        let tokenData = null;
+
+        for (let i = 0; i < check.recordset.length; i++){
+            let found = await bcrypt.compare(refreshToken, check.recordset[i].TOKEN);
+            if (found) {
+                tokenData = check.recordset[i];
+                break;
+            }
+        }
+        if (!tokenData) {
+            return res.status(404).send({ message: "Không tìm thấy refresh token" });
+        }
+        // if (check.recordset.length === 0) {
+        //     return res.status(404).send({ message: "Không tìm thấy token" });
+        // }
+        // let tokenData = check.recordset[0];
+        if (new Date(tokenData.EXPIRESAT) < new Date()) {
+            return res.status(401).send({ message: 'Refresh token đã hết hạn! Vui lòng đăng nhập lại!' });
```

Comment:
[Must]
Cả 2 trường hợp `access token hết hạn` và `refresh token hết hạn` đều trả về status code 401. Vậy nên cần trả thêm một field nào đó để phía FE biết cụ thể là lỗi nào để có hướng xử lý đúng:
- access token hết hạn -> dùng refresh để lấy token mới
- refresh token hết hạn -> quay về trang đăng nhập để lấy access + refresh token mới.
Ví dụ
```suggestion
            return res.status(401).json({ errorCode: 'invalid_refresh_token'  message: 'Refresh token đã hết hạn! Vui lòng đăng nhập lại!' });
```